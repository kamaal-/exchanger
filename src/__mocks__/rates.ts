export const rates = {
    "EUR": {
        currency: 'EUR',
        rate: 0.9096697899
    },
    "SEK": {
        currency: 'SEK',
        rate: 9.9072136814
    },
    "USD": {
        currency: 'USD',
        rate: 1.0
    },
    "GBP": {
        currency: 'GBP',
        rate: 0.8110161012
    }
};

export const rate2 = {
    "EUR": {
        currency: 'EUR',
        rate: 1.9096697899
    },
    "SEK": {
        currency: 'SEK',
        rate: 9.9072136814
    },
    "USD": {
        currency: 'USD',
        rate: 1.0
    },
    "GBP": {
        currency: 'GBP',
        rate: 0.8110161012
    }
};