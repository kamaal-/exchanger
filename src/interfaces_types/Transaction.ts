
export default interface Transaction{
    detail:string;
    wallets: {
        from: {
            currency: string;
            symbol: string;
        };
        to: {
            currency: string;
            symbol: string;
        };
    };
    from:{
        before: string;
        diff: string;
        after: string;
    }
    to: {
        before: string;
        diff: string;
        after: string;
    }

    exchange: string;
    time: number;
}