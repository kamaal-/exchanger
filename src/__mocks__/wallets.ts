import IWallet from "../interfaces_types/Wallet";
import { rates } from "./rates";
import Wallet from "../services/models/Wallet";
import { symbol } from "prop-types";

export const wallets: IWallet[] = [
  {
    currency: rates["EUR"].currency,
    rate: rates["EUR"].rate,
    balance: 10.23,
    symbol: "€"
  },
  {
    currency: rates["USD"].currency,
    rate: rates["USD"].rate,
    balance: 150.45,
    symbol: "$"
  },
  {
    currency: rates["GBP"].currency,
    rate: rates["GBP"].rate,
    balance: 320.4,
    symbol: "l"
  }
];

export const balances = {
  fromWallet: {
    currency: "EUR",
    balance: 10.23,
    diff: 0,
    label: "Balance"
  },
  toWallet: {
    currency: "USD",
    balance: 150.45,
    diff: 0,
    label: "Balance"
  }
};

export const balances_ok = {
  fromWallet: {
    currency: "EUR",
    balance: 10.23,
    diff: -3.23,
    label: "Balance"
  },
  toWallet: {
    currency: "USD",
    balance: 150.45,
    diff: 4.56,
    label: "Balance"
  }
};

export const balances_not_ok = {
  fromWallet: {
    currency: "EUR",
    balance: 10.23,
    diff: -10.24,
    label: "Balance"
  },
  toWallet: {
    currency: "USD",
    balance: 150.45,
    diff: 4.56,
    label: "Balance"
  }
};


export const transactions = [ 
  { 
     "detail":"You have transfered 34 kr to RM.",
     "wallets":{ 
        "from":{ 
           "currency":"SEK",
           "symbol":"kr"
        },
        "to":{ 
           "currency":"MYR",
           "symbol":"RM"
        }
     },
     "from":{ 
        "before":"100 kr",
        "diff":"34 kr",
        "after":"66 kr"
     },
     "to":{ 
        "before":"200 RM",
        "diff":"14.47 RM",
        "after":"214.47 RM"
     },
     "exchange":"1 kr = 0.42 RM",
     "time":1570979215945
  },
  { 
     "detail":"You have transfered 2 € to £.",
     "wallets":{ 
        "from":{ 
           "currency":"EUR",
           "symbol":"€"
        },
        "to":{ 
           "currency":"GBP",
           "symbol":"£"
        }
     },
     "from":{ 
        "before":"95 €",
        "diff":"2 €",
        "after":"93 €"
     },
     "to":{ 
        "before":"104.38 £",
        "diff":"1.75 £",
        "after":"106.13 £"
     },
     "exchange":"1 € = 0.87 £",
     "time":1570979187093
  },
  { 
     "detail":"You have transfered 5 € to £.",
     "wallets":{ 
        "from":{ 
           "currency":"EUR",
           "symbol":"€"
        },
        "to":{ 
           "currency":"GBP",
           "symbol":"£"
        }
     },
     "from":{ 
        "before":"100 €",
        "diff":"5 €",
        "after":"95 €"
     },
     "to":{ 
        "before":"100 £",
        "diff":"4.38 £",
        "after":"104.38 £"
     },
     "exchange":"1 € = 0.87 £",
     "time":1570973116868
  }
];