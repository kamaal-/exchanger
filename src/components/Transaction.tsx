import React from "react";
import { useSpring } from "react-spring";
import {
  WalletWrapper,
  CurrencyLogo,
  DoubleLogo,
  WalletBalance,
  WalletRow,
  TransactionDetail
} from "../theme/index";
import { Size } from "../interfaces_types/types";
import ITransaction from "../interfaces_types/Transaction";
import moment from "moment";

type Props = {
  transaction: ITransaction;
};

const Transaction = ({ transaction }: Props) => {
  const styleProps = useSpring({
    from: { opacity: "0" },
    to: { opacity: "1" }
  });
  return (
    <WalletWrapper style={styleProps}>
      <WalletRow style={{ alignItems: "flex-start" }}>
        <DoubleLogo>
          <CurrencyLogo 
            data-testid={`transation-currency`}
            style={{ position: "absolute", left: "0", top: "0" }}
            currency={transaction.wallets.from.currency}
            size={Size.Medium}
          >
            {transaction.wallets.from.symbol}
          </CurrencyLogo>
          <CurrencyLogo
            data-testid={`transation-currency-2`}
            style={{ position: "absolute", left: "25px", top: "0" }}
            currency={transaction.wallets.to.currency}
            size={Size.Medium}
          >
            {transaction.wallets.to.symbol}
          </CurrencyLogo>
        </DoubleLogo>
        <TransactionDetail
          data-testid={`transation-details`}
          style={{
            marginTop: "0px",
            marginLeft: "10px",
            textAlign: "left",
            marginRight: "20px"
          }}
        >
          {transaction.detail}
        </TransactionDetail>
        <WalletBalance
          data-testid={`transation-balance`}
          style={{
            marginLeft: "auto",
            marginTop: "0px",
            textAlign: "right",
            minWidth: "80px"
          }}
          size={Size.Small}
        >
          {transaction.exchange}
        </WalletBalance>
      </WalletRow>
      <WalletRow style={{ marginTop: "10px" }}>
        <WalletBalance size={Size.Small} data-testid={`transation-history-from`}>
          {transaction.from.before} &#8594; {transaction.from.diff} &#8594;{" "}
          {transaction.from.after}
        </WalletBalance>
      </WalletRow>
      <WalletRow>
        <WalletBalance size={Size.Small} data-testid={`transation-history-to`}>
          {transaction.to.before} &#8594; {transaction.to.diff} &#8594;{" "}
          {transaction.to.after}
        </WalletBalance>
        <WalletBalance style={{ marginLeft: "auto" }} size={Size.Small} data-testid={`transation-date`}>
          {moment(transaction.time).format("DD/MM/YYYY HH:mm")}
        </WalletBalance>
      </WalletRow>
    </WalletWrapper>
  );
};

export default Transaction;
