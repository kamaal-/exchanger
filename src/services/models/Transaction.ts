import ITransaction from '../../interfaces_types/Transaction'
import IWallet from '../../interfaces_types/Wallet'
import {BalancePair, Balance} from '../../interfaces_types/types'
import moment from 'moment'
import {formateDecimalPlaces} from '../../utils' 

export default class Transaction implements ITransaction {
    detail = '';
    wallets = {
        from: {
            currency: '',
            symbol: ''
        },
        to: {
            currency: '',
            symbol: ''
        },
    };
    from = {
        before: '',
        diff: '',
        after: ''
    };
    to = {
        before: '',
        diff: '',
        after: ''
    }
    exchange = '';
    time = moment().valueOf()

    constructor(fromWallet?:IWallet, toWallet?:IWallet, balances?: BalancePair){
        if(fromWallet && toWallet && balances){
            const fromBalance:Balance = balances['fromWallet'];
            const toBalance:Balance = balances['toWallet'];
            debugger
            this.wallets.from = {currency: fromWallet.currency, symbol: fromWallet.symbol};
            this.wallets.to = {currency: toWallet.currency, symbol: toWallet.symbol};

            this.from.before = `${formateDecimalPlaces(fromWallet.balance)} ${fromWallet.symbol}`;
            this.from.diff = `${formateDecimalPlaces(fromBalance.diff)} ${fromWallet.symbol}`;
            this.from.after = `${formateDecimalPlaces(fromBalance.balance)} ${fromWallet.symbol}`;

            this.to.before = `${formateDecimalPlaces(toWallet.balance)} ${toWallet.symbol}`;
            this.to.diff = `${formateDecimalPlaces(toBalance.diff)} ${toWallet.symbol}`;
            this.to.after = `${formateDecimalPlaces(toBalance.balance)} ${toWallet.symbol}`;
            this.exchange = fromWallet.convert ?  `1 ${fromWallet.symbol} = ${fromWallet.convert(1, toWallet)} ${toWallet.symbol}` : '';
            this.detail = `You have transfered ${formateDecimalPlaces(fromBalance.diff)} ${fromWallet.symbol} to ${toWallet.symbol}.`;
        }
    }

    convenientConsttructor = (data:ITransaction) => {
        this.detail = data.detail;
        this.from = data.from;
        this.exchange = data.exchange;
        this.to = data.to;
        this.wallets = data.wallets;
        this.time = data.time;
    }
}