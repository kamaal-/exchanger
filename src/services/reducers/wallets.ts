import { WalletsState, Action } from "../../interfaces_types/types";
import { WALLETS, CONVERT, APP } from "../action_types/index";

export const initialState: WalletsState = {
  wallets: null,
  fromWallet: null,
  toWallet: null,
  balances: null,
  leftWallet: null,
  rightWallet: null,
  transactions: [],
  value: ""
};

export default (state = initialState, { type, payload }: Action) => {
  switch (type) {
    case WALLETS.ADD:
      return { ...state, wallets: payload };
    case WALLETS.REMOVE:
      return { ...state, wallets: payload };
    case WALLETS.FROM:
      return { ...state, fromWallet: payload };
    case WALLETS.TO:
      return { ...state, toWallet: payload };
    case WALLETS.SWITCH:
      return {
        ...state,
        toWallet: state.fromWallet,
        fromWallet: state.toWallet
      };
    case CONVERT.DO:
      return { ...state };
    case WALLETS.LEFT:
      return { ...state, leftWallet: payload };
    case WALLETS.RIGHT:
      return { ...state, rightWallet: payload };
    case WALLETS.RESET:
      return {
        ...state,
        balances: payload.balances,
        fromWallet: payload.fromWallet,
        toWallet: payload.toWallet
      };
    case CONVERT.RUN:
      return { ...state, balances: payload };
    case WALLETS.ADD_TRANSACTION:
      return { ...state, transactions: payload };
    case WALLETS.SET_VALUE:
      return { ...state, value: payload };
    case APP.SUCCESS_POLLING:
    case APP.START_POLLING:
      return { ...state };
    default:
      return initialState;
  }
};
