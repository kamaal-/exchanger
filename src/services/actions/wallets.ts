import { WALLETS, CONVERT } from "../action_types/";
import { ThunkDispatch } from "redux-thunk";
import { saveInLocalStorage } from "../../utils/";
import Wallet from "../../interfaces_types/Wallet";
import { StoreState, Balance, BalancePair } from "../../interfaces_types/types";
import Transaction from "../models/Transaction";
import ITransaction from "../../interfaces_types/Transaction";

export const setFromWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  let { balances } = getState().Wallets;
  balances = balances ? { ...balances } : {};
  await saveInLocalStorage("fromWallet", wallet);
  dispatch({ type: WALLETS.FROM, payload: wallet });
  balances["fromWallet"] = {
    currency: wallet.currency,
    balance: wallet.balance,
    diff: 0,
    label: "Balance"
  };
  dispatch({ type: CONVERT.RUN, payload: balances });
};

export const setLeftWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  await saveInLocalStorage("leftWallet", wallet);
  dispatch({ type: WALLETS.LEFT, payload: wallet });
};

export const setToWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  let { balances } = getState().Wallets;
  balances = balances ? { ...balances } : {};
  await saveInLocalStorage("toWallet", wallet);
  dispatch({ type: WALLETS.TO, payload: wallet });
  balances["toWallet"] = {
    currency: wallet.currency,
    balance: wallet.balance,
    diff: 0,
    label: "After"
  };
  dispatch({ type: CONVERT.RUN, payload: balances });
};

export const setRightWallet = (wallet: Wallet) => async (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  await saveInLocalStorage("rightWallet", wallet);
  dispatch({ type: WALLETS.RIGHT, payload: wallet });
};

export const switcWallets = () => (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  let { fromWallet, toWallet } = getState().Wallets;
  if (fromWallet && toWallet) {
    dispatch(setToWallet(fromWallet));
    dispatch(setFromWallet(toWallet));
    dispatch(triggerConversion("0", toWallet, fromWallet));
  }
};

export const resetWallets = (fromWallet: Wallet, toWallet: Wallet) => (
  dispatch: ThunkDispatch<{}, {}, any>
) => {
  if (fromWallet && toWallet) {
    const balances = {
      fromWallet: {
        currency: fromWallet.currency,
        balance: fromWallet.balance,
        diff: 0,
        label: "Balance"
      },
      toWallet: {
        currency: toWallet.currency,
        balance: toWallet.balance,
        diff: 0,
        label: "Balance"
      }
    };

    dispatch({
      type: WALLETS.RESET,
      payload: { balances, fromWallet, toWallet }
    });
    dispatch(setValue(""));
  }
};

export const saveTransaction = (
  transaction: ITransaction,
  key = "transactions"
) => (dispatch: ThunkDispatch<{}, {}, any>) => {
  if (transaction && transaction instanceof Transaction) {
    let transactionsList: ITransaction[] = [];
    try {
      const oldTransactions = localStorage.getItem(key);
      if (oldTransactions) {
        transactionsList = JSON.parse(oldTransactions);
      }
      transactionsList = transactionsList.map(t => {
        const obj = new Transaction();
        obj.convenientConsttructor(t);
        return obj;
      });
      transactionsList.unshift(transaction);
      saveInLocalStorage(key, transactionsList);
      dispatch({ type: WALLETS.ADD_TRANSACTION, payload: transactionsList });
    } catch (error) {
      transactionsList.unshift(transaction);
      saveInLocalStorage(key, transactionsList);
      dispatch({ type: WALLETS.ADD_TRANSACTION, payload: transactionsList });
    }
  }
};

export const triggerConversion = (
  input: string,
  fromWallet: Wallet,
  toWallet: Wallet
) => (dispatch: ThunkDispatch<{}, {}, any>) => {
  if (input && fromWallet && toWallet) {
    dispatch({ type: CONVERT.DO, payload: { input, fromWallet, toWallet } });
  }
};

export const getTransaction = (
  fromWallet: Wallet,
  toWallet: Wallet,
  balances: BalancePair
) => new Transaction(fromWallet, toWallet, balances);

export const exchange = (cb = () => {}) => (
  dispatch: ThunkDispatch<{}, {}, any>,
  getState: () => StoreState
) => {
  const state: StoreState = getState();
  const { fromWallet, toWallet, balances, wallets } = state.Wallets;

  // detuct from wallet
  if (fromWallet && toWallet && balances) {
    const fromBalance: Balance = balances["fromWallet"];
    const toBalance = balances["toWallet"];
    // only pass the exchange if their any values and value can deduct from purse.
    if (
      fromBalance.diff &&
      toBalance.diff &&
      Math.abs(fromBalance.diff) <= fromWallet.balance
    ) {
      // keep the transaction before update wallets
      const transaction = getTransaction(fromWallet, toWallet, balances);

      fromWallet.balance = fromBalance.balance;
      toWallet.balance = toBalance.balance;
      const newWallets = (wallets || [])
        .map(w => (w.currency === fromWallet.currency ? fromWallet : w))
        .map(w => (w.currency === toWallet.currency ? toWallet : w));
      dispatch(updateWallets(newWallets));
      dispatch(setFromWallet(fromWallet));
      dispatch(setToWallet(toWallet));
      const newBalances = {
        fromWallet: {
          currency: fromWallet.currency,
          balance: fromWallet.balance,
          diff: 0,
          label: "Balance"
        },
        toWallet: {
          currency: toWallet.currency,
          balance: toWallet.balance,
          diff: 0,
          label: "Balance"
        }
      };
      dispatch({
        type: WALLETS.RESET,
        payload: { balances: newBalances, fromWallet, toWallet }
      });
      dispatch(saveTransaction(transaction));
      dispatch(resetWallets(fromWallet, toWallet));
      dispatch(setValue(""));
    }
  }
};

export const updateWallets = (wallets: Wallet[]) => async (
  dispatch: ThunkDispatch<{}, {}, any>
) => {
  if (wallets && wallets.length >= 0) {
    await saveInLocalStorage("wallets", wallets);
    dispatch({ type: WALLETS.ADD, payload: wallets });
  }
};

export const setValue = (value: string) => async (
  dispatch: ThunkDispatch<{}, {}, any>
) => {
  dispatch({ type: WALLETS.SET_VALUE, payload: value });
};
