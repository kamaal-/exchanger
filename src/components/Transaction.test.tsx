import React from 'react'
import Transaction from './Transaction'
import { render } from '@testing-library/react'
import {transactions} from '../__mocks__/wallets'

describe('Transaction component', () => {

    it('should match snapshot', () => {
        const { asFragment } = render(<Transaction transaction={transactions[0]}/>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders with currency logo from', () => {
        const { getByTestId } = render(<Transaction transaction={transactions[0]}/>)
        expect(getByTestId(`transation-currency`).textContent).toBe('kr')
    });

    it('renders with currency logo to', () => {
        const { getByTestId } = render(<Transaction transaction={transactions[0]}/>)
        expect(getByTestId(`transation-currency-2`).textContent).toBe('RM')
    });

    it('renders with details', () => {
        const { getByTestId } = render(<Transaction transaction={transactions[0]}/>)
        expect(getByTestId(`transation-details`).textContent).toBe('You have transfered 34 kr to RM.')
    });

    it('renders with exchange', () => {
        const { getByTestId } = render(<Transaction transaction={transactions[0]}/>)
        expect(getByTestId(`transation-balance`).textContent).toBe('1 kr = 0.42 RM')
    });

    it('renders with exchange history from', () => {
        const { getByTestId } = render(<Transaction transaction={transactions[0]}/>)
        expect(getByTestId(`transation-history-from`).textContent).toBe('100 kr → 34 kr → 66 kr')
    });

    it('renders with exchange history to', () => {
        const { getByTestId } = render(<Transaction transaction={transactions[0]}/>)
        expect(getByTestId(`transation-history-to`).textContent).toBe('200 RM → 14.47 RM → 214.47 RM')
    });

    it('renders with exchange date', () => {
        const { getByTestId } = render(<Transaction transaction={transactions[0]}/>)
        expect(getByTestId(`transation-date`).textContent).toBe('13/10/2019 19:06')
    });
})