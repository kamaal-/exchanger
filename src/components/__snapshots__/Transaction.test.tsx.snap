// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa cRcnPO"
    style="opacity: 0;"
  >
    <div
      class="sc-dnqmqq dOtUEt"
      style="align-items: flex-start;"
    >
      <div
        class="sc-gZMcBi lnOzAh"
      >
        <div
          class="sc-bZQynM gKtDJe"
          data-testid="transation-currency"
          style="position: absolute; left: 0px; top: 0px;"
        >
          kr
        </div>
        <div
          class="sc-bZQynM ebwyKT"
          data-testid="transation-currency-2"
          style="position: absolute; left: 25px; top: 0px;"
        >
          RM
        </div>
      </div>
      <div
        class="sc-gqjmRU ezkMFm"
        data-testid="transation-details"
        style="margin-top: 0px; margin-left: 10px; text-align: left; margin-right: 20px;"
      >
        You have transfered 34 kr to RM.
      </div>
      <div
        class="sc-htpNat soCIS"
        data-testid="transation-balance"
        style="margin-left: auto; margin-top: 0px; text-align: right; min-width: 80px;"
      >
        1 kr = 0.42 RM
      </div>
    </div>
    <div
      class="sc-dnqmqq dOtUEt"
      style="margin-top: 10px;"
    >
      <div
        class="sc-htpNat soCIS"
        data-testid="transation-history-from"
      >
        100 kr → 34 kr → 66 kr
      </div>
    </div>
    <div
      class="sc-dnqmqq dOtUEt"
    >
      <div
        class="sc-htpNat soCIS"
        data-testid="transation-history-to"
      >
        200 RM → 14.47 RM → 214.47 RM
      </div>
      <div
        class="sc-htpNat soCIS"
        data-testid="transation-date"
        style="margin-left: auto;"
      >
        13/10/2019 19:06
      </div>
    </div>
  </div>
</DocumentFragment>
`;
